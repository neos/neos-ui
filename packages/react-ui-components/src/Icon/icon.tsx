import React, {PureComponent} from 'react';

import {FontAwesomeIconProps} from '@fortawesome/react-fontawesome';
import {PickDefaultProps, Omit} from '../../types';
import FontAwesomeIcon from './fontAwesomeIcon';
import ResourceIcon, {ResourceIconProps} from './resourceIcon';

type IconSize = 'xs' | 'sm' | 'lg' | '2x' | '3x';
type IconPadding = 'none' | 'left' | 'right';
type IconColor = 'default' | 'warn' | 'error' | 'primaryBlue';

export interface IconTheme {
    readonly icon: string;
    readonly 'icon--huge': string;
    readonly 'icon--large': string;
    readonly 'icon--big': string;
    readonly 'icon--small': string;
    readonly 'icon--tiny': string;
    readonly 'icon--paddedLeft': string;
    readonly 'icon--paddedRight': string;
    readonly 'icon--spin': string;
    readonly 'icon--color-warn': string;
    readonly 'icon--color-error': string;
    readonly 'icon--color-primaryBlue': string;
}

export interface IconProps extends Omit<FontAwesomeIconProps, 'icon'> {
    /**
     * We use the react component FortAwesome provides to render icons.
     * we will pass down all props to the component via {...rest} to expose it's api
     * https://github.com/FortAwesome/react-fontawesome/blob/master/README.md
     */

    /**
     * The identifier of the icon to render.
     * Can be a font-awesome icon identifier or a asset resource.
     */
    readonly icon?: string;

    /**
     * The (accessibility) label for this icon
     */
    readonly label?: string;

    /**
     * Controls the padding around the icon in a standardized way.
     */
    readonly padded?: IconPadding;

    /**
     * An optional `className` to attach to the wrapper.
     */
    readonly className?: string;

    /**
     *  Adjust the color of the icon
     */
    readonly color?: IconColor;

    /**
     *  Adjust the size of the icon
     */
    readonly size?: IconSize;

    /**
     *  An optional css theme to be injected.
     */
    readonly theme?: IconTheme;
}

type DefaultProps = PickDefaultProps<IconProps, 'color' | 'padded' | 'size'>;

export const defaultProps: DefaultProps = {
    color: 'default',
    padded: 'none',
    size: 'sm'
};

class Icon extends PureComponent<IconProps> {
    public static readonly defaultProps = defaultProps;

    public render(): JSX.Element |Â null {
        const {icon} = this.props;

        if (icon && icon.substr(0, 11) === 'resource://') {
            return <ResourceIcon {...this.props as ResourceIconProps} />;
        } else {
            return <FontAwesomeIcon {...this.props} />;
        }
    }
}

export default Icon;
