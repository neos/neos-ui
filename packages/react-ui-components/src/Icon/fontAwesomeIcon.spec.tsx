import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import {defaultProps, IconProps} from './icon';
import FontAwesomeIcon from './fontAwesomeIcon';

describe('<FontAwesomeIcon/>', () => {
    const props: IconProps = {
        ...defaultProps,
        icon: 'fooIconClassName',
        theme: {
            'icon': 'iconClassName',
            'icon--big': 'bigClassName',
            'icon--color-error': 'color-errorClassName',
            'icon--color-primaryBlue': 'color-primaryBlueClassName',
            'icon--color-warn': 'color-warnClassName',
            'icon--paddedLeft': 'paddedLeftClassName',
            'icon--paddedRight': 'paddedRightClassName',
            'icon--small': 'smallClassName',
            'icon--spin': 'spinClassName',
            'icon--tiny': 'tinyClassName',
            'icon--large': 'largeClassName',
            'icon--huge': 'hugeClassName'
        }
    };

    it('should render correctly.', () => {
        const wrapper = shallow(<FontAwesomeIcon {...props}/>);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should allow the propagation of "className" with the "className" prop.', () => {
        const wrapper = shallow(<FontAwesomeIcon {...props} className="fooClassName"/>);

        expect(wrapper.prop('className')).toContain('fooClassName');
    });

    it('should call the "fontAwesome.getClassName" api method and render the returned className.', () => {
        const wrapper = shallow(<FontAwesomeIcon {...props}/>);
        expect(wrapper.hasClass('fooIconClassName')).toBeTruthy();
    });

    it('should allow the propagation of custom "icon" with the "icon" prop.', () => {
        const wrapper = shallow(<FontAwesomeIcon {...props} icon="bazIconClassName"/>);
        expect(wrapper.hasClass('bazIconClassName')).toBeTruthy();
    });
});

