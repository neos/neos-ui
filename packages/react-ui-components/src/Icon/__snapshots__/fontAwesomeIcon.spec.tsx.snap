// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FontAwesomeIcon/> should render correctly. 1`] = `
<FontAwesomeIcon
  border={false}
  className="iconClassName fooIconClassName"
  fixedWidth={false}
  flip={null}
  icon={
    Array [
      "fas",
      "fooIconClassName",
    ]
  }
  inverse={false}
  listItem={false}
  mask={null}
  pull={null}
  pulse={false}
  rotation={null}
  size="sm"
  spin={false}
  swapOpacity={false}
  symbol={false}
  title=""
  transform={null}
/>
`;
